File: app/src/main/java/com/orionedwards/mtglifecounter/FloatingView.java
package com.orionedwards.mtglifecounter;

import android.animation.Animator;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

import java.util.concurrent.Callable;

import static com.orionedwards.mtglifecounter.Util.pxToDp;

public class FloatingView extends RelativeLayout {
    public static final long
            DEFAULT_CALLBACK_MILLIS = 2200,
            DEFAULT_PAUSE_MILLIS = 1400,
            DEFAULT_FADE_MILLIS = 600;

    public interface BeforeShowCallback {
        void beforeShow(long callbackDurationMillis);
    }

    final @NonNull View mInnerView;

    private @Nullable BeforeShowCallback mBeforeShow;
    private @Nullable Runnable mBeforePause;

    public FloatingView(Context context, @NonNull View innerView, Float cornerRadius) {
        super(context);

        mInnerView = innerView;

        float padding = pxToDp(context, 8);
        setPadding((int)padding, (int)padding, (int)padding, (int)padding);

        LayoutParams p = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        p.addRule(CENTER_IN_PARENT, TRUE);
        addView(innerView, p);

        float minVal = pxToDp(context, 40);
        setMinimumWidth((int)minVal); // try to be a square
        setMinimumHeight((int)minVal);

        // setup rounded corners
        GradientDrawable bg = new GradientDrawable();
        bg.setCornerRadius(pxToDp(context, cornerRadius));
        bg.setColor(Color.BLUE);
        setBackground(bg);
    }

    public void setBeforeShow(BeforeShowCallback value) {
        mBeforeShow = value;
    }

    public void setBeforePause(Runnable value) {
        mBeforePause = value;
    }

    @TargetApi(21) // lollipop, android 5.0 needed for setElevation only
    private void setElevationIfAvailable() {
        if(Build.VERSION.SDK_INT >= 21) {
            setElevation(4);
        }
    }

    public void showInView(@NonNull RelativeLayout parent, @NonNull RelativeLayout.LayoutParams layoutParams) {
        showInView(parent, layoutParams, DEFAULT_CALLBACK_MILLIS, DEFAULT_PAUSE_MILLIS, DEFAULT_FADE_MILLIS);
    }

    public void showInView(@NonNull final RelativeLayout parent, @NonNull final RelativeLayout.LayoutParams layoutParams, long callbackDurationMillis, long pauseDurationMillis, long fadeDurationMillis) {
        if(getParent() != null) {
            return; // already parented, animation mustn't have finished or something
        }

        parent.addView(this, layoutParams);
        setAlpha(1f); // doesn't animate, it just pops in
        setElevationIfAvailable();

        if(mBeforeShow != null) {
            mBeforeShow.beforeShow(callbackDurationMillis);
        }

        if(mBeforePause != null){
            Util.delay(callbackDurationMillis, mBeforePause);
        }

        final FloatingView self = this;
        this.animate()
                .alpha(0)
                .setStartDelay(callbackDurationMillis + pauseDurationMillis)
                .withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        parent.removeView(self);
                    }
                })
                .start();
    }

    public void remove() {
        removeView(mInnerView);
        ((ViewGroup)getParent()).removeView(this);
    }
}

File: app/src/main/java/com/orionedwards/mtglifecounter/FloatingView.kt
package com.orionedwards.mtglifecounter

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import com.orionedwards.mtglifecounter.Util.delay
import com.orionedwards.mtglifecounter.Util.pxToDp

class FloatingView(
        context: Context,
        private val innerView: View,
        cornerRadius: Float?) : RelativeLayout(context)
{

    var beforeShow: ((callbackDurationMillis: Long) -> Unit)? = null
    var beforePause: (() -> Unit)? = null

    // standard "context only" constructor to satisfy lint
    constructor(context: Context) : this(context, View(context), null)

    init {
        val padding = pxToDp(context, 8f)
        setPadding(padding.toInt(), padding.toInt(), padding.toInt(), padding.toInt())

        val p = RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT)
        p.addRule(RelativeLayout.CENTER_IN_PARENT, RelativeLayout.TRUE)
        addView(innerView, p)

        val minVal = pxToDp(context, 40f)
        minimumWidth = minVal.toInt() // try to be a square
        minimumHeight = minVal.toInt()

        // setup rounded corners
        val bg = GradientDrawable()
        bg.cornerRadius = pxToDp(context, cornerRadius!!)
        bg.setColor(Color.BLUE)
        background = bg
    }

    override fun setBackgroundColor(color: Int) {
        // custom background drawable for rounded cornders
        val bg = background as GradientDrawable
        bg.setColor(color)
    }

    fun pulse() {
        this.animate()
                .scaleX(1.3f)
                .scaleY(1.3f)
                .setDuration(200)
                .start()
    }

    @TargetApi(21) // lollipop, android 5.0 needed for setElevation only
    private fun setElevationIfAvailable() {
        if (Build.VERSION.SDK_INT >= 21) {
            elevation = 20f
        }
    }

    @JvmOverloads
    fun showInView(parent: RelativeLayout, layoutParams: LayoutParams, callbackDurationMillis: Long = DEFAULT_CALLBACK_MILLIS, pauseDurationMillis: Long = DEFAULT_PAUSE_MILLIS) {
        if (getParent() != null) {
            return  // already parented, animation mustn't have finished or something
        }

        parent.addView(this, layoutParams)
        alpha = 1f // doesn't animate, it just pops in
        setElevationIfAvailable()

        beforeShow?.invoke(callbackDurationMillis)

        beforePause?.let{
            delay(callbackDurationMillis, Runnable(it))
        }

        val self = this
        this.animate()
                .alpha(0f)
                .setStartDelay(callbackDurationMillis + pauseDurationMillis)
                .withEndAction { parent.removeView(self) }
                .start()
    }

    fun remove() {
        removeView(innerView)
        val parent = parent as ViewGroup
        parent.removeView(this)
    }

    companion object {
        const val DEFAULT_CALLBACK_MILLIS: Long = 2200
        const val DEFAULT_PAUSE_MILLIS: Long = 1400
    }
}
